[SYCL][ESIMD] Add test on logical not operator

vasilytric opened the pull request.

vasilytric committed:
[SYCL][ESIMD] Add test on logical not operator

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_logical_not.cpp:
```diff
@@ -0,0 +1,154 @@
+    shared_vector<int> logical_not_res_is_correct(1,
+                                                  shared_allocator<int>(queue));
+    shared_vector<int> logical_not_res_type_is_correct(
+        1, shared_allocator<int>(queue));
```
```suggestion
    shared_vector<int> is_correct_result(1,
                                                  shared_allocator<int>(queue));
    shared_vector<int> is_correct_type(
        1, shared_allocator<int>(queue));
```
1) We are in `operator_logical_not.cpp` anyway. So in scope of this test there is no other operator.
2) You have `TestCaseT::get_description()` later, but use explicit `logical not operator` sometimes

vasilytric replied:
these variables was renamed [here](https://github.com/intel/llvm-test-suite/pull/784/commits/e868d819250aebcd33b836e14dd71cd62fb4e257#diff-eac54e76a314e24c2d279609017dedef6dab795a4af3a589567c94d08070bf55R87-R88)

yuriykoch replied:
You are right, the `operator_result` fits much better

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_logical_not.cpp:
```diff
@@ -0,0 +1,154 @@
+                             FlagT *const logical_not_res_is_correct) {
+    auto simd_obj = simd<DataT, NumElems>();
+    simd_obj.copy_from(ref_data);
+    const auto logical_not_result = !simd_obj;
```
That's how I see it: `logical_not_operator::call_simd_ctor::logical_not_result` - why not just `logical_not_operator::call_simd_ctor::operator_result` instead?

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/784/commits/e868d819250aebcd33b836e14dd71cd62fb4e257#diff-eac54e76a314e24c2d279609017dedef6dab795a4af3a589567c94d08070bf55L33-L46)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_logical_not.cpp:
```diff
@@ -0,0 +1,154 @@
+  static std::string get_description() { return "logical not"; }
+
+  template <typename DataT, int NumElems, typename FlagT>
+  static bool call_simd_ctor(const DataT *const ref_data, DataT *const out,
```
You don't call constructor here. Again, why not just `TestCaseT::run`?

vasilytric replied:
I think that `TestCaseT::run` not so self descriptable, I suggest use `TestCaseT::call_operator`, I've done it [here](https://github.com/intel/llvm-test-suite/pull/784/commits/e868d819250aebcd33b836e14dd71cd62fb4e257#diff-eac54e76a314e24c2d279609017dedef6dab795a4af3a589567c94d08070bf55R98-R100)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/operator_logical_not.cpp:
```diff
@@ -0,0 +1,154 @@
+  const auto u_types = get_tested_types<tested_types::uint>();
+  const auto s_types = get_tested_types<tested_types::sint>();
+  const auto dims = get_all_dimensions();
+  const auto context = unnamed_type_pack<logical_not_operator>::generate();
```
1) Once again, that's not a `constructor context` we had for tests on constructors. It's more a test case now, or a specific operator to use.
2) You might just pass it as a template custom parameter to the `for_all_combinations`, like
```
using operator = logical_not_operator;
for_all_combinations<run_test, operator>(u_types, dims, queue)`
...
// where
template <typename TestCaseT, typename DataT, typename DimT> class run_test
```
3) Looks like you are going to move out `run_test` to the common file, to use over multiple operators. That's why you have a separate class for `logical_not_operator`. Am I right to suppose it? Why are you stopping halfway then, with no header file provided? It would be a great way to show why `struct logical_not_operator` is required.

vasilytric replied:
It's a good advice, I've applied it [here](https://github.com/intel/llvm-test-suite/pull/784/commits/e868d819250aebcd33b836e14dd71cd62fb4e257#diff-eac54e76a314e24c2d279609017dedef6dab795a4af3a589567c94d08070bf55R145-R148)

vasilytric committed:
Merge branch 'intel' into operator_logical_not

vasilytric committed:
[SYCL][ESIMD] Change test logic and small updates

- Compare values from simd_mask on host side
- Rename variables
- Move test into operators folder

vasilytric committed:
[SYCL][ESIMD] Add debug messages

vasilytric committed:
[SYCL][ESIMD] Remove assignment

vasilytric committed:
[SYCL][ESIMD] Remove type verification

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_logical_not.cpp:
```diff
@@ -0,0 +1,161 @@
+  template <typename DataT, int NumElems, typename OperatorResultT>
+  static bool call_operator(const DataT *const ref_data, DataT *const out,
+                            OperatorResultT *const operator_result) {
+    auto simd_obj = simd<DataT, NumElems>();
```
```suggestion
    simd<DataT, NumElems> simd_obj;
```

vasilytric replied:
this comment has been applied in this commit: 7da36ff78106ec5529ce117f6facc3defad83c3e

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_logical_not.cpp:
```diff
@@ -0,0 +1,161 @@
+                        const std::string &data_type) {
+    assert(ref_data.size() == NumElems &&
+           "Reference data size is not equal to the simd vector length.");
+    std::cout << "in run_verification" << std::endl;
```
Just a friendly reminder, I'm sure you are aware. Here and elsewhere

vasilytric replied:
thanks for the reminder

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_logical_not.cpp:
```diff
@@ -0,0 +1,161 @@
+using namespace sycl::ext::intel::experimental::esimd;
+using namespace esimd_test::api::functional;
+
+// Descriptor class for the case of calling logical not operator.
+struct logical_not_operator {
```
Would you mind to comment why do we need a separate descriptor class in this specific case?
Maybe some dependency, or architecture considerations, or some TODO for the future.

Without it I have a feeling that we could remove this class and move it contents into the `run_verification`

vasilytric replied:
I suggest do not do this change because we can add some other cases to this test by addition new classes like `logical_not_operator `, but if your comment will be applied we couldn't do it

vasilytric committed:
[SYCL][ESIMD] Disable test due to it freezed on Linux OS

vasilytric committed:
[SYCL][ESIMD] Revert all changes that has done in the scope of debugging

vasilytric committed:
[SYCL][ESIMD] Use var decl context instead of initializer

vasilytric committed:
[SYCL][ESIMD] Disable test on simd logical not operator

vasilytric committed:
[SYCL][ESIMD] Enable test only for Windows OS

vasilytric commented:
@kbobrovs Would you mind to review this PR?
